# Vector_DB : [ 1."Cosince Similarity search" ]

# Hugging_Face :

- If I will select the Model - 
- Lets say - I search for the -> Emdedding & select the Model this one -> [Qwen/Qwen3-Embedding-8B ]
- URL : https://huggingface.co/Qwen/Qwen3-Embedding-8B
- 8B - 8 Billiong Parameter.
- Inside this, there is code-> we have to download the model from the "Hugging_Face". Then we have to pass the document.

# Note :
- Below this all we can find inside we can see-> documents = [
- "The capital of China is Beijing.",
- "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement 
- of planets around the sun.", ]
- This will convert our document into a Vector Space. Then we can go ahead with the Similarity search & Everywhere we can find out the
- exact kind of story.

- # Note :
- Compute the (cosine) similarity between the query and document embeddings
- similarity = model.similarity(query_embeddings, document_embeddings)
- print(similarity)

- I can go ahead and use this model, but what if get better. & Again, I dont have to go and download the model.

# Euri API : 

- Bettter option, we have ["Euri_API"]. Follow stpes :
- Go to "Secources" section
- Go to the - Code_Example : next "Text Embedding and Encoding".
- Code Below is the inside Euri one. Here we can see this line of code : [ "model": "text-embedding-3-small" ]
- This model Access has been given to, this model is itself coming from whom, this is from "OpenAI" itself.
- This total length of Context it generate is - (1536). which we will discussed later.
- This is best model -> [ "model": "text-embedding-3-small" ], we will be find out which is having the all of the knowledge.
- This is the problem, which we have discussed, the major problem is not to do a similarity search, is very small mathematics.
- The Major problem is to use the best of the best model. Which understand the data, base on those understanding data, it convert my data - into the Vector or Numerical space.

- Here incase of the "Euri API", here below code already provided, Text Embedding small model 3 is already provided.
- We dont have to do any kind of Installation & we dont have to do any kind of setup & we can directly call the API.

- Just Copy and paste code from this Euron-> ["Euri_API"]. 

- Follow this step to get below Code, inside the "Euri".
- Go to the - "Sources" section, 
- Next go to the -> Code_Example : 
- Next "Text Embedding and Encoding". <- Here we will find the code inside the "Euri_API", Just copy and paste it here below.

# How will I get my token? 

- Again simple I can go to, Follow this steps were I will get this token.
- Go to the "Euri API"\
- Next go to the "API key" and create a new API Key or it will craete the Token, Just copy and paste or replace in the Str section (****).
- For example : Lets say, I have create one API,inside my Euri -> ["euri-9f10184195d5280594a3124f6b1dc07bcfb7331da8c9c30b2c3575307079538e"]

# Note : Don't run Below code 

- Now Here Below is the code, Don't run next to this code run.

In [ ]:

import requests
import numpy as np

def generate_embeddings(text):
    url = "https://api.euron.one/api/v1/euri/embeddings"
    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer ********************"
    }
    payload = {
        "input": text,
        "model": "text-embedding-3-small"
    }

    response = requests.post(url, headers=headers, json=payload)
    data = response.json()
    
    embedding = np.array(data['data'][0]['embedding'])
    
    return embedding

text = "The weather is sunny today."

embedding = generate_embeddings(text)

# Lets Start :

- Below I code I will going to custimized and make some changes in it.
- Authorization": "Bearer ***************" <- Here I have to replace this "API" Key that, I have create just now inside the ["Euri API"].
- Now, This is complete code -> [ "Authorization": "Bearer euri-9f10184195d5280594a3124f6b1dc07bcfb7331da8c9c30b2c3575307079538e" ]

- 1.So, Below we have the function -> def generate_embeddings(text):
- 2.we have the replace API key
- "Payload" wise, we are using this trying to use this particular model -> [ "model": "text-embedding-3-small" ]. Text Embedding free from
- OpenAI itself.

# What data we are trying to pass?

- The data we are trying to pass is whether 'Sunny today" or just replace this data with my data.
-  this section data replace with my new data.
- This is the data -> [ "text = "The weather is sunny today."] <- I am replacing it with 
- my new data -> ["My name is Vijay and I am a Data Scientist"].

- If I will going to execute this, I am going to get some sort of a output.


In [1]:
import requests
import numpy as np

def generate_embeddings(text):
    url = "https://api.euron.one/api/v1/euri/embeddings"
    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer euri-9f10184195d5280594a3124f6b1dc07bcfb7331da8c9c30b2c3575307079538e"
    }
    payload = {
        "input": text,
        "model": "text-embedding-3-small"
    }

    response = requests.post(url, headers=headers, json=payload)
    data = response.json()
    
    embedding = np.array(data['data'][0]['embedding'])
    
    return embedding

text = "My name is Vijay and I am Data Scientist"

embedding = generate_embeddings(text)

- Once Execute it now lets print the "Embedding"
- As a outcome, we can see this is basically the Numerical representation.

- Here, below we are basically are able to do the conversation, this conversation simpley means that, I am trying to represent
- This data -> text = [ "My name is Vijay and I am Data Scientist" ], in a ( 1536 ) dimension.
- So, any model that we are going to use in this entire world, to convert our data into the numerical representation.
- It will always be having a fix dimension, It will always try to give a return into a fix dimension.

In [2]:
embedding

array([-0.01132254, -0.03332492, -0.0133588 , ...,  0.03207554,
        0.03726529,  0.00630699])

# To check What is the Length of this Vector?

- As a outcome, we can see that is 1536, its a the  shape = 1536.
- So the Length is ( 1536 ), means whatever data we are going to pass.
- lets say we pass this -> text = "My name is Vijay and I am Data Scientist" inside my code. whether its a single word or whether its
- a hundread of words, or maybe 1,000 of words, in a one single line, whatever data we are going to pass inside this model -
- this one ->  [ "model": "text-embedding-3-small" ], eventually it will try to convert those data into the numerical values of
- the lenth of -> ( 1536 ) or in other words, I can say that dimension, means it is trying to ( 1536 ) different different coordinates
- just like the X coordinate and Y coordinate or Z coordinate. 
- So, ( 1536 ) coordinate represent to the one point. That one point in that spaces, is representing a data.

In [3]:
len(embedding)

1536

# Note : 

- Continue.....

- Open Notebook, Because made a note, I am draw some diagram and explained this.

# Note : Now lets continue same thing.

- Incase of different different model, the sizes are going to be different. Maybe 700 something, maybe 1,000 or maybe 1,500 something.
- So, OpenAI model, which we are using inside the "UriAPI". It is going to give us return of ( 1536 ) & 
- the meaning of this ( 1536 ) is representing this particular data, -> text = "My name is Vijay and I am Data Scientist"
- This outcome we can see over here -> array([-0.01132321, -0.03330284, -0.01334757, ...,  0.03207741,
-      0.03726746,  0.0063434 ])
- This is a one single dot(.) only, for example ->   
- dot(.) [ (a,b,c,d,e,f,g,h) ] etc, in a single dot(.)
- This entire dataset which is trying to give me as an output, which is just a single dot(.) in these many dimentions.

# Note : 1D,2D, 3D, 4D or 1336 D.

- Lets say there is 4 co-ordinate, we can't visualized in programing language, Max to Max till the 3D we will be able to visualized it.
- But, there is possibility that 4th co-ordinate is exists.
- 1.I can use lets say - (X,Y,G & A). to represent a 4 coordinate.
- 2.May be (X,Y,G,A & B ) to represent a 5 co-ordinate.
- 3.may be (X,Y,G,A,B & C ) to represent a 6 co-ordinate.
- This line outcome, we can see -> array([-0.01132321, -0.03330284, -0.01334757, ...,  0.03207741,
-         0.03726746,  0.0063434 ]) <- this is what  I am trying to do over here.
- I am just trying to represent this data ->  text = [ "My name is Vijay and I am Data Scientist" ] into the ( 1536 ) co-ordinate.
- This is what it means. This is the one of the vector.
- What is the dimension of this vector ?
- The dimention of this vector is -> [ 1536 ]
- So, we have -> 1D, 2D, 3D, 4D similary we have here 1536D.

# Public Question ?

- What is the region choosing the (1356) ?
- The ( 1356 ) is a model output dimension. So, whenever we talk about the Generative AI, we always talk about the "Context Length".
- What is the meaning of "Content Length" ?
- The meaning is kind of output, the length of the output, it will be able to give me or length of the input it will be able to take.
- Because whatever model, we are going to create, eventually its going to be a "Neural Network", its a Neurons.
- There is transformer architecture but it is going to be a Neruron.
- So, every model is having a constant. Constant input dimension & every model having the Costant output dimension.

- We simply say that ok fine, recent Model ChatGPT has released they have 128k of Context, means 1,28,000 of a context length,means
- that is the output it will be able to yield in a single go. 
- For one single input, it will be give me that kind of or maybe 4k of context or maybe 8k of context.
- So, similarily for the Embedding model, whoever has build this particular model, they try to set that, ok fine. this is going to
- give you this much of output. -> ( 1536) is an ouput, we will be able to find out a model which is going to give us 
- maybe be 700 dimension & maybe 1,028 Dimension as an output.

# Note : 

- So, we just saw that with the help of "Euri_API" how we can pass any kind of data & eventually I will be able to generate the
- embedding. 

# Hugging_Face : Continue........

- Now coming back to hugging face.
- URL : https://huggingface.co/Qwen/Qwen3-Embedding-8B
- Open this Url and scole down and go to the -> [ Qwen3 Embedding Series Model list  ]

- Inside the Hugging face, we cant just find out the one model, but also a thousand of model, we will be able to find out.
- Even we can try to use this model - URL : [ https://huggingface.co/Qwen/Qwen3-Embedding-8B ] <- inside this 
- Go to here -> [ Qwen3 Embedding Series Model list  ] <- even inside this, we an try to use this models, to perform the to this 
- embedding operation, so here 

 # No.1 : 

- 1.Model Type -> Text_Embedding
- 2.Models -> Qwen3-Embedding-0.6B  <- Here 0.6 Billion parameter model, we are going to use.
- 3.Size -> 0.6B  <- This is the size of the model, 0.6 Billion parameter model, what this means.
- 4.Layer -> 28 <- Layer inside this model means transformer layer is basically 28, a stacked layer and 
- 5.Sequence length : This sequence length is 32k.
- 6.Embedding Dimension : Embedding Dimension is 1024, the model that we are using, what was the dimension? it was ( 1536 ).
- SO, what dimension this model is going to give, interms of embedding? it is 1024. whatever model that we are going to use in this entire
- world,all of these model comes with the specification. All this specification people are going to mention. Does not matter from where
- we are going to use from where, but yeah Embedding Dimension, is 1024.


# No.2 :
- Now, similarily 

- 1.Model Type -> Text_Embedding
- 2.Models -> "Qwen3-Embedding-4B" <- so, Qwen3 series embedding 4 Billion parameter, if we are going to use. The embedding dimention
-  is ( 2560 ) <- So the final output or final dimension, it is going to give. But, our on our case it was ( 1356 ).


# No.3 :
- Model Type : Text Embedding
- Models : Qwen3-Embedding-8B  <- if we are going to use the Qwen3 series embedding 8 Billion parameters, so it will going to give us 
- embedding dimension is equals to  ( 4096 ).

- So, inside the QWEN series, we will be able to find out, these 3 models. 

# Note :

- Now lets suppose I have to use one of these model.

- click on this No.1.
- Lets use this one -> Text Embedding | Qwen3-Embedding-8B | URL : [ https://huggingface.co/Qwen/Qwen3-Embedding-0.6B  ]

# Lets use the smallest one ( 0.6B ) or 0.6 Billion parameter.

- come to this section scroll down ad inside this URL this one -> ( #Sentence Transformers Usage ).
- So, here we can find the code -> ( #from sentence_transformers import SentenceTransformer ) <- here where we are going to create,
- a sentences transformer & then we are going to create an
-  Object sentences transformer -> ( #model = SentenceTransformer("Qwen/Qwen/Qwen3-Embedding-0.6B") by passing this 
- model name ("Qwen/Qwen/Qwen3-Embedding-0.6B"), whatever model we are going to use, then we are going to pass our data,


- This one -> we can see this inside this document, below code I just copy paste.
-# The queries and documents to embed
- queries = [
-    "What is the capital of China?",
-    "Explain gravity",
- ]
- documents = [
-    "The capital of China is Beijing.",
-    "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for "
- the movement of planets around the sun.",]

- & it will going to take our data & then it is going to encoded it.

# Note :

- Just copy code and past below is the same above I mention. Like "Qwen" I can use any of these models.

- This is Url where we can find this code : [  https://huggingface.co/Qwen/Qwen3-Embedding-0.6B ]
- I have pasted code below, but I am going to clean this remove unnecessary this from here.

# Note : Don't execute below code,
- just copy and paste. we need to remove the things and once ready then execute it.

In [ ]:
# Requires transformers>=4.51.0
# Requires sentence-transformers>=2.7.0

from sentence_transformers import SentenceTransformer

# Load the model
model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B")

# We recommend enabling flash_attention_2 for better acceleration and memory saving,
# together with setting `padding_side` to "left":
# model = SentenceTransformer(
#     "Qwen/Qwen3-Embedding-0.6B",
#     model_kwargs={"attn_implementation": "flash_attention_2", "device_map": "auto"},
#     tokenizer_kwargs={"padding_side": "left"},
# )

# The queries and documents to embed
queries = [
    "What is the capital of China?",
    "Explain gravity",
]
documents = [
    "The capital of China is Beijing.",
    "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun.",
]

# Encode the queries and documents. Note that queries benefit from using a prompt
# Here we use the prompt called "query" stored under `model.prompts`, but you can
# also pass your own prompt via the `prompt` argument
query_embeddings = model.encode(queries, prompt_name="query")
document_embeddings = model.encode(documents)

# Compute the (cosine) similarity between the query and document embeddings
similarity = model.similarity(query_embeddings, document_embeddings)
print(similarity)
# tensor([[0.7646, 0.1414],
#         [0.1355, 0.6000]])


# Note :  

- Here I am, not looking for the similarity search, I am looking for the -> ( document_embeddings ).
- Now my code is ready after remove everything.
- Once we execude this below code it will start downloading, the those model 1st and once it download it, 
- I am just calling the -> "model.encode" & I am trying to pass my -> "documents", this particular document, I am trying to pass.
- I am trying to pass my data.

- Before execute below code, we need to install one Library -> [ sentence_transformers ] -> pip install sentence_transformers

In [7]:
pip install sentence_transformers

Defaulting to user installation because normal site-packages is not writeable
   ---------------------------------------- 0.0/12.0 MB ? eta -:--:--
   - -------------------------------------- 0.5/12.0 MB 2.5 MB/s eta 0:00:05
   ----- ---------------------------------- 1.6/12.0 MB 3.8 MB/s eta 0:00:03
   --------- ------------------------------ 2.9/12.0 MB 4.8 MB/s eta 0:00:02
   -------------- ------------------------- 4.5/12.0 MB 5.5 MB/s eta 0:00:02
   ------------------- -------------------- 5.8/12.0 MB 5.8 MB/s eta 0:00:02
   ----------------------- ---------------- 7.1/12.0 MB 5.9 MB/s eta 0:00:01
   --------------------------- ------------ 8.4/12.0 MB 5.9 MB/s eta 0:00:01
   --------------------------------- ------ 10.0/12.0 MB 6.0 MB/s eta 0:00:01
   -------------------------------------- - 11.5/12.0 MB 6.1 MB/s eta 0:00:01
   ---------------------------------------- 12.0/12.0 MB 6.0 MB/s eta 0:00:00
   ---------------------------------------- 0.0/564.3 kB ? eta -:--:--
   -----

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


- So, here we understanding how we are going to use the some of the open Sources model, which is widely avaiable to all of us.
- If I can try to use one of any model, I can use or else there is thousand of model, if I want to use then I can use it.

# Note :

- In Hugging_face they always give us the sample code, everytime we can find. Again Sample code is nothing everytime, we can able to 
- find out that, they are trying to call a "Sentences_Transformer" & they try to pass the model name -> ("Qwen/Qwen3-Embedding-0.6B")
- which will be given eventually inside a Hugging_face itself, we just have to copy this name -> ("Qwen/Qwen3-Embedding-0.6B")
- all the time. we have to copy this model name, then hit the "Hugging_Face", then it will try to download everything, then it will do
- the inferencing with those models.

In [4]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B")

documents = [
    "The capital of China is Beijing.",
    "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun.",
]

document_embeddings = model.encode(documents)


# Note :

- Above code I run inside my local, 1st it will try to download the model -> model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B")
- then it will try to perform this operation in my local ->
-  documents = [
-   "The capital of China is Beijing.",
-    "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the -movement of planets around the sun.",
- ]

- Now everyting looks good and we are able to download it, now lets call my "document_embeddings".
- As a outcome, we can see it has returing me 2 embedding, because inside document we can see there is 2 paragraph sentences.
- 1st sentences -> 1."The capital of China is Beijing.", & 
- 2nd sentences -> 2."Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is - responsible for the movement - of planets around the sun.",

- But, for then both of the sentences, it is going to give it to me the similar dimension.

In [5]:
document_embeddings 

array([[-0.04718245, -0.02091702,  0.00366519, ...,  0.05624256,
         0.07076066, -0.0171463 ],
       [-0.05312933, -0.01502102, -0.00126578, ...,  0.0036741 ,
        -0.0205426 ,  0.01957481]], dtype=float32)

# Continue :  1:10:13 

- I can even try to check what is the dimension.
- So, maybe document Embedding of [0th], so lets try to extract very 1st one.
- As a outcome as we can see over here.

In [6]:
document_embeddings[0]

array([-0.04718245, -0.02091702,  0.00366519, ...,  0.05624256,
        0.07076066, -0.0171463 ], dtype=float32)

- I can even try to check the Length of it. what is the length ?
- As a outcome, we can see that Length is 1024.
- now go and check inside the document & we can see (1024) <- this is what they have mention.
- In document mention Qwen 3 embedding 0.6 Billion Parameter model -> ("Qwen/Qwen3-Embedding-0.6B")
- This is going to going to give us Embedding Dimension.A vector dimension is ( 1024 ) & yes as an outcome I am able to get the ( 1024 ).

In [7]:
len(document_embeddings[0])

1024

- I can even able to use the "Open Sources model", A model which is like widely avaiable over a "Hugging_Face", with out even an API.
- But the thing is that, they "Hugging face" download the model, inside my local machine.
- So, if the model size will keep inceasing, everytime we have to download those, big models and then we have to do an inferencing. 
- or maybe we have to host the same model on a cloude infrastructure. SO that without even using a "Euri AP"I".
- some of can't able to use the Software, we are ok with that. Even a using the "Euri API".
- I am able to convert my data into a vector spaces, we can do it. 

# Note: What we have discussed ?

- 1.we are able to convert our dataset into a numerical spaces, into a vector spaces by using the "Euri API".
- 2.& 2nd is without using the "Euri API"

- Let's say, "Euri API" is not avaiable to me. Not an issue, at all its compltely fine.
- So, I have a tons of model is avaiable outside. I can try to use those models and I can try to convert my dataset into a vector spaces.
- That is sometjng which we can try to do it. 

- So conversation of the data part is done, as per our discussion, we are now able to convert, Now we in the situation where we can 
- try to convert our data, into this kind of structure : 

- 1.D1 = [1,0,1]
- 2.D2 = [0,1,1]
- 3.D3 = [1,1,0]
- 4.D4 = [1,0,0]

- a vector space i which we will be able to convert it. 


# Note :  Cosince Similarity : In a notebook I have already there just check this.

- How we will be able to prove that, ok we have multiple sentences and whta is the similarity between between these 2 sentences in a 
- Practical manner. SO how we will be able to understand it.
- To do that lets craete some sort of Builups.

- Lets support, I will going to use those model above one same. Again we will try to understand with "Euri API".

- Lets support, we have this sentences below :
- I am going to write my own sentences, so the my very 1st sentences is 
- 1st sentences -> [ "my name is vijay & I am a Data Scientist" ]. 
- This is my very 1st sentences.
- 2nd sentences -> [ "vijay is my name vijay and I am a Data Scientist" ]
- 3rd sentences -> [ "Appple is red and it is a fruit" ]
- 4th Sentences -> [ "babana is yellow and it is a fruit" ]
- 5th sentences -> ["Vijay is giving a class for GenAI and going to start class for GenAI interview preparation" ]

- This is some random sentences, I am trying to write over here.
- so, these are the multiple sentences.In total I have taken 5 sentences, now I will try to pass all of these sentences inside
- this one -> (documents)  <- inside this particular model. So it is going to provide me 5 vectors.
- We are just trying to co-related my theory with my practical -  D1,D2, D3, D4 
- 1.D1 = [1,0,1]
- 2.D2 = [0,1,1]
- 3.D3 = [1,1,0]
- 4.D4 = [1,0,0]

- Similary, we ahve vector for that, so we have written 5 sentences. Execute it and go and explain.

In [8]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B")

documents = [
    "my name is vijay & I am a Data Scientist",
    "vijay is my name vijay and I am a Data Scientist",
    "Appple is red and it is a fruit",
    "babana is yellow and it is a fruit",
    "Vijay is giving a class for GenAI and going to start class for GenAI interview preparation",]

document_embeddings = model.encode(documents)

- Now for this 5 sentences, I will go and check, these 5 vectors.
- As a outcome, we can see over here, 
- Total we have 5 vectors, we are trying to use a "Qwen3" model 0.6 Billion parameter model. we had basically below we can see outcome,
- we have total 5 vectors, 
- Vector No.1, vector No.2, Vector No.3, vector No.4, & Vector No.5

# 1:22:20 Continueeee

In [9]:
document_embeddings

array([[ 0.00844296, -0.03947532, -0.01244079, ..., -0.02714348,
         0.02418726, -0.02235063],
       [ 0.01451294, -0.03753193, -0.01256607, ..., -0.03329413,
         0.03206141, -0.01124465],
       [-0.06847467,  0.06475923, -0.00253857, ..., -0.00598949,
         0.0145832 ,  0.01016517],
       [-0.04552928,  0.02589087, -0.00620311, ..., -0.01881102,
        -0.03197147,  0.01239174],
       [ 0.01821066, -0.0271492 , -0.01146782, ...,  0.01832489,
         0.03078685, -0.02545705]], dtype=float32)

# Next :

- Now I have to check, vector No.1, technically a [0th] index vector :
- This 1st one is -> 1. [ 0.00844296, -0.03947532, -0.01244079, ..., -0.02714348, 0.02418726, -0.02235063] is representing,
- This one -> [ "my name is vijay & I am a Data Scientist", ]

- This 2nd one is -> [ 0.01451294, -0.03753193, -0.01256607, ..., -0.03329413, 0.03206141, -0.01124465], is representing,
- This one -> [ "vijay is my name vijay and I am a Data Scientist",]
- Like this till rest 5 vectors represent.


# Note :

- Now, I have to check what is the similarity, between sentences No.1 and sentences No.2, means 

- 1.what is similarty between -> "my name is vijay & I am a Data Scientist", &  "Appple is red and it is a fruit",?
- 2.what is similarty between -> "vijay is my name vijay and I am a Data Scientist", & "babana is yellow and it is a fruit",?
- We have to check, I can try to use maybe a Euclidean distance or I can try to use the Cosine similarity

# Next : Cosine similarity

- let's go ahead and do the Cosine similarity these 2 vectors.
- I have decided & what is the similarity between this sentences No.1 -> ["my name is vijay & I am a Data Scientist"] &
- sentences No.2 -> ["vijay is my name vijay and I am a Data Scientist"]
- Because interms of English, it is looks like there is alot of Similarity.
- But we will try to check the Mathematically as well, what is the diffence between these 2 vectors that we have.

# Library : 

- To find out this Cosine similarity, we can try to use a direct Library called as-> from sklearn.metrics.pairwise import cosine_similarity
- Here, we will be find out direct function as a -> ("cosine_similarity").
- If we would like to use this one use it, if we dont want then we know the mathematics. 
- Manually try to calculate the dot(.) product of vector, 2 vector divided by magnitude of one vector into Magnitude of the 2nd vector. 
- either way we can try to go that is completely fine. But yes inside Python inside Sklearn there is a function which is a they have 
- already provided to us. we can try to pass the Embedding, & we will be able to find out the "Cosine Similarity" 

# Note :

- Whether we manual or formula wise? 

- Lets go with the manual wise.

- So, from Numpy import there is basically dot(.), in a numerator we have a dot(.) product.
- So, vector A dot(.) vector B.
- So, we can try to import a dot(.), which is going to build my Numerator & If I have to build my denominator.
- I have to do a Normilization this one and this one -> ||A|| x ||B|| & do the multiplication.
- This is called the Normilization -> ||A||, finding the magnitude is called as norm or normilization.
- code -> [ "from numpy.linalg import norm"] - so Numpy then linalg and norm.

In [10]:
from numpy import dot
from numpy.linalg import norm

# Next : Now choose a vector, 

- 1.sentences 1 or sent1 is this one -> [ 0.00844296, -0.03947532, -0.01244079, ..., -0.02714348, 0.02418726, -0.02235063],

- from the document_emdedding try to get [0]th indexes element. 

- 2.sentences 2 or sent2 is this one -> [ 0.01451294, -0.03753193, -0.01256607, ..., -0.03329413, 0.03206141, -0.01124465],

- sentence 2 is nothing but document_embedding [1] indexs element.

In [11]:
sent1 = document_embeddings[0]
sent2 = document_embeddings[1]

# Find similarity between this 2 sentences.

- Now I have to find out the what is the similarity between Sentence 1 or sent1 and Sentences 2 or sent2 , 
- with the help of cosine similarity, we know the formula, the fomula. The formula for the Cosine similarity is 
- dot(.) of sentences numerator sentence 1 & sentences 2 / divided by normalization of sentences 1 into sentence 2 execute.
- As a outcome, we can see that, the similarity is ( 0.97 ) % which means, it is closer to (1). Closer to (1) means 
- there is high similarity.

In [12]:
dot(sent1, sent2) / (norm(sent1) * norm(sent2))

np.float32(0.97699696)

# Next : Similarity check between (Sentence No.1 and Sentences No.3)

- 1."my name is vijay & I am a Data Scientist", & 3."Appple is red and it is a fruit".
- As we can see the outcome is that, the similarity between sentences 1 or Sent1 and Sentences 2 or Sent2 is (0.37)%.
- It is very less, so it is closer to Zero, if it was above 50+ then it comes under near to (1). But unfortunately it comes under (0).
- closer to Zero means, its a "orthogonal"."Orthogonal" means 19 degree perpendicular to each other, in the mathematical term.
- The 2 vectors are perpendicular to each other.

In [13]:
sent1 = document_embeddings[0]
sent3 = document_embeddings[2]

In [14]:
dot(sent1, sent3) / (norm(sent1) * norm(sent3))

np.float32(0.3774256)

# Next : Similarity check between Sentence No.3 and Sentences No.4 

- sentences 3 or sent3 : "Appple is red and it is a fruit", & sentences 4 or sent4 : "babana is yellow and it is a fruit",
- there are similarity, fruits both are. As we can see it is giving me (0.75) % which means near to [1].

- This is how I can convert my data into, vector space & mow I am able to do the similarity check as well.

In [15]:
sent3 = document_embeddings[2]
sent4 = document_embeddings[3]

In [16]:
dot(sent3, sent4) / (norm(sent3) * norm(sent4))

np.float32(0.7524843)

# Note :

- Now If I am going to use "Euri API", technically i will end up getting vector. Once I have the vector, once my dataset is avaiable
- into this form -> 

- [ 0.00844296, -0.03947532, -0.01244079, ..., -0.02714348,0.02418726, -0.02235063],
- [ 0.01451294, -0.03753193, -0.01256607, ..., -0.03329413, 0.03206141, -0.01124465],

- Then I will do the math. for example this way - [ dot(sent3, sent4) / (norm(sent3) * norm(sent4)) ]
- to check how similarity dataset is with respect to all the other dataset.

# Note : 

- Similar way we will more in forwarding classes, whenever we try to covere topic "Langchain" | "Lama Index" | "Lang Graph" | "CrewAI"
- whenever we try to doing a Tool calling, & we end up doing a "Agent" building | Maybe "RAG" Application. So everytime we are going to 
- use the somehow to use this technique to convert my data, into a vector spaces & then process it. 

# NOte : 

- The main Agenda of todays class was a ["Vector database"].
- what vector database will do? we are going to store this data -> 
- [ 0.00844296, -0.03947532, -0.01244079, ..., -0.02714348,0.02418726, -0.02235063],
- [ 0.01451294, -0.03753193, -0.01256607, ..., -0.03329413, 0.03206141, -0.01124465], etc......
- The data that we are able to produce, incase of SQL system we try to create a tables, in tables we try to insert the data.
- Incase of NoSQL the document, technically a "JSON" format is called the "Documents" a Key Value pair.
- we try to store it onside the NoSQL database.

- Incase of vector Database, we are goigng to store this -> -

-  [ 0.00844296, -0.03947532, -0.01244079, ..., -0.02714348,0.02418726, -0.02235063],
- [ 0.01451294, -0.03753193, -0.01256607, ..., -0.03329413, 0.03206141, -0.01124465],

- Then whenever, we are going to search, lets say I have alreay store this data.--->


- documents = [
-   "my name is vijay & I am a Data Scientist",
-    "vijay is my name vijay and I am a Data Scientist",
-   "Appple is red and it is a fruit",
-    "babana is yellow and it is a fruit",
-  "Vijay is giving a class for GenAI and going to start class for GenAI interview preparation",]

# Note :  "Query Vector""

# Lets say we have a ["Query Vector"].

- Query says that -> ["who is Vijay"]<- , Based on the Query we have to pull a data from where, from the VectorDB

- From here -> this line code : 
-documents = [
-    "my name is vijay & I am a Data Scientist",
-    "vijay is my name vijay and I am a Data Scientist",
-    "Appple is red and it is a fruit",
-    "babana is yellow and it is a fruit",
-    "Vijay is giving a class for GenAI and going to start class for GenAI interview preparation",]

- This is the data which, I have store inside vector BD, so how based on the Query, I will be able to pull the data?
- Cosine Similarity, so what I will do is, I will convert this query, - > q = ["who is Vijay"] <-

- Let's imagine that, this data -> this line code outcome : [document_embeddings] :

- array([[ 0.00844296, -0.03947532, -0.01244079, ..., -0.02714348,
-        0.02418726, -0.02235063],
-       [ 0.01451294, -0.03753193, -0.01256607, ..., -0.03329413,
-         0.03206141, -0.01124465],
-       [-0.06847467,  0.06475923, -0.00253857, ..., -0.00598949,
-         0.0145832 ,  0.01016517],
- This above data is avaiable in my "Vector DB", This was my raw dataset.

# Next : 

- Now user has fired a query, this is query -> ["who is Vijay"].
- Bases on the query, we have to extract the relevant output, which is avaiable inside the "VectorDB"
- 1st of all, I will try to convert my query -> ["who is Vijay"] into a Embeddings,by using the exact same method which we have done above
- I will try to convert my query, "Q" into a embeddings.
- As a outcome, if we call this -> ( query_embed ) <- we can see this is my embedding.

- array([[ 0.02358586, -0.03157804, -0.01153745, ...,  0.00884003,
-        0.02084586, -0.01900646]], dtype=float32)

- Now, I have a Embedding for a Query

# Next : 

- Then, I have so many documents : this line code -> document_embeddings


- array([[ 0.00844296, -0.03947532, -0.01244079, ..., -0.02714348,
-         0.02418726, -0.02235063],
-       [ 0.01451294, -0.03753193, -0.01256607, ..., -0.03329413,
-         0.03206141, -0.01124465],
-       [-0.06847467,  0.06475923, -0.00253857, ..., -0.00598949,
-        0.0145832 ,  0.01016517],
-       [-0.04552928,  0.02589087, -0.00620311, ..., -0.01881102,
-        -0.03197147,  0.01239174],
-       [ 0.01821066, -0.0271492 , -0.01146782, ...,  0.01832489,
-         0.03078685, -0.02545705]], dtype=float32)

- There are so many documents are out there, means all above this dataset is avaiable in my database.

- So, what I will do, with respect to this below query, this embedding of the query -> 

- array([[ 0.02358586, -0.03157804, -0.01153745, ...,  0.00884003,
-         0.02084586, -0.01900646]], dtype=float32)

- I will try to check with 1st, 2nd, 3rd, 4th & with 5th like this. 
- I will try to check the distance, a similarity -> with 1st, 2nd, 3rd, 4th & with 5th like this. I will check similarity of all of it.
- Whichever will give me more similarity, that will going to be my answer. That will be the answer of that particular query.

In [19]:
q = ["who is Vijay"]

In [20]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B")

query_embed = model.encode(q)

In [21]:
query_embed

array([[ 0.02358586, -0.03157804, -0.01153745, ...,  0.00884003,
         0.02084586, -0.01900646]], dtype=float32)

# Calculate Distance 

- Lets suppose I am going to create, "Distance", between this 2 queries. 
- Distance or Dist with sentence 1,2,3,4,5 lets say, with total 4 distance I am calculation these 4 sentences.
- We have 5 sentences right but we will do with 4 now because count start from 1 to 4 not from 0 to 5. sentence (0) skipped. 
- As we can see below, with these 4 sentence, I am able to calculated the distance. 

In [23]:
dist1 = dot(query_embed, sent1) / (norm(query_embed) * norm(sent1))

In [24]:
dist2 = dot(query_embed, sent2) / (norm(query_embed) * norm(sent2))

In [25]:
dist3 = dot(query_embed, sent3) / (norm(query_embed) * norm(sent3))

In [26]:
dist4 = dot(query_embed, sent4) / (norm(query_embed) * norm(sent4))

# print all together

- Now, try to print this distance, maybe try to print all together -> Dist1,Dist2,Dist3,Dist4
- As a outcome, we can see over here, I am able to get the answer.

In [28]:
dist1,dist2,dist3,dist4

(array([0.65222585], dtype=float32),
 array([0.66339064], dtype=float32),
 array([0.36987156], dtype=float32),
 array([0.37711608], dtype=float32))

# Note : 

- So, here I had a query, what is my query this one -> q = ["who is Vijay"]
- which sentences is going to be top answer for my query. 
- This one right -> [ array([0.66339064] ], why ? because its 0.66 % which is closer to (1). & this is for distance 2 sentences 2 & 
- what was the sentences2 or sent2  by the way this one -> ["vijay is my name vijay and I am a Data Scientist",]
- This sentences is closer to (1), which we can see in our outcome (0.66%). 
- So, this sent2 is closer to the query sentences that, we have called as -> ["who is Vijay"]
- This is where can go and use a DB or Database. 
- So, we are storing our embedding into a databases, whenever there will be a query, user is going to fire a query, it will go a and
- it will try to check that which is similar,it is eventually going to give the answer.
- This is something which is going to happened internally inside our database or DB again & again. 
- Whenever, we are going to fire a query, we are going to store a data. 

# Realtime Use casses :
- what we want to create.

- Now lets come to the use case wise, what will be the realtime casses?
- For example we all are working on a company or lets say if you are student and trying to prepare a Notes.
- Lets say working on company, we have alot of HR document, Finance document, maybe project document we all keep use to alot of documents.
- while studying or while working. I wanted to create a database, so where I can store all of those documents. whatever data is inside the
- the documents, and I would like to create a chat interface, where it should answer or where it should always give me an answer.
- based on the existing knowledge or existing data based not from the Global answer, I am not looking answer direct from ChatGPT. 
- What I am expecting is that it should give me answer from my data. Because ChatGPT or other AI model will not having my data, it doesnot
- have the access or idea about the private data, what is going inside our Finance depart or HR department? what kind of notes we have
- taken, ChatGpt will not have those idea, and we need an answer from that.
- No problem read all of the PDF document and whatever files that we have convert each and everything into a embeddings. 
- Store into a Vector DB. In a chat interface whenever we are going to chat it is going to do a similarity search & then it will try to
- bring a dataset for me.


# Note : Vector Database : Next Class

- Tomorrow class, what we will do is we will discussed about the some of the Vector databases -> 

- 1.Facebook or Meta AI similarity search 
- 2.Chroma DB 
- 3.Pinecone DB
- 4.Waviet DB
- 5.Qdrant

- we will discuused this and some of the vector DB, we will be able to install on our Local and some of these Vector DB already 
- providing us a Cloude interface. So, we dont have to do any kind of a local system installation. Just go ahead and store data there
- itself after converting into the Embedding. 
- Now a days even a MongoDB has started giving a vector DB, Now a days every vendor is providing.
- Now a days most of every vendor is providing a "Vector DB".
- But, above mention list are very famous one. 

# Note :

- So, we have discussed so far basic of Embedding. What technically Embedding is and how will be perform those embedding
- with the help of the "OpenAI" model which is avaiable inside URI, or otherwise we can try to use any other model from the "Higging Face"
- then we can try to convert those data into a "Embedding Vector".
- Tomorrow class, will discussed how we can do embedding in practical manner, by using some of these databases.

# For Example : 

- 1.Facebook AI Similarity Search
- 2.Chroma DB 
- 3.Qdrant DB 
- 4.Pinecone DB etc other DB also we will discussed which is avaiable in the market.